hazelcast-sql-core/src/main/java/com/hazelcast/sql/impl/calcite/opt/physical/index/IndexResolver.java

```diff
@@ -135,21 +159,122 @@ private IndexResolver() {
         }
 
-        List<RelNode> rels = new ArrayList<>(supportedIndexes.size());
 
         for (MapTableIndex index : supportedIndexes) {
             // Create index scan based on candidates, if possible. Candidates could be merged into more complex
             // filters whenever possible.
-            RelNode rel = createIndexScan(scan, distribution, index, conjunctions, candidates);
 
-            if (rel != null) {
-                rels.add(rel);
+            RelNode relAscending = createIndexScan(scan, distribution, index, conjunctions, candidates, false);
+            RelNode relDescending = createIndexScan(scan, distribution, index, conjunctions, candidates, true);
+
+            if (relAscending != null) {
+                rels.add(relAscending);
+            }
+
+            if (relDescending != null) {
+                rels.add(relDescending);
             }
         }
 
         return rels;
     }
 
+
+    /**
+     * Filters out index scans which collation is covered (prefix based) by another index scan in the rels.
+     *
+     * @param rels the list of index scans
+     * @return the list of filtered out index scans
+     */
+    @SuppressWarnings("checkstyle:NPathComplexity")
+    private static List<RelNode> excludeCoveredCollations(List<RelNode> rels) {
+        // Order the index scans based on their collation
+        TreeMap<RelCollation, RelNode> relsTreeMap = new TreeMap<>((coll1, coll2) -> {
+            // Compare the collations field by field
+            int coll1Size = coll1.getFieldCollations().size();
+            int coll2Size = coll2.getFieldCollations().size();
+
+            for (int i = 0; i < coll1Size; ++i) {
+                if (i >= coll2Size) {
+                    // The coll1 has more fields and the prefixes are equal
+                    return 1;
+                }
+
+                RelFieldCollation fieldColl1 = coll1.getFieldCollations().get(i);
+                RelFieldCollation fieldColl2 = coll2.getFieldCollations().get(i);
+                // First, compare directions
+                int cmp = fieldColl1.getDirection().compareTo(fieldColl2.getDirection());
+                if (cmp == 0) {
+                    // Directions are the same
+                    if (fieldColl1.getFieldIndex() == fieldColl2.getFieldIndex()) {
+                        // And fieldIndex is the same, try the next field
+                        continue;
+                    } else {
+                        return Integer.compare(fieldColl1.getFieldIndex(), fieldColl2.getFieldIndex());
+                    }
+                }
+                return cmp;
+            }
+
+            // All the fields from coll1 are equal to the fields from coll2, compare the size
+            return Integer.compare(coll1Size, coll2Size);
+        });
+
+        // Put the rels into the ordered TreeMap
+        for (RelNode rel : rels) {
+            relsTreeMap.put(rel.getTraitSet().getTrait(RelCollationTraitDef.INSTANCE), rel);
```

Comment:
What if there are two relations with the same collation? E.g. a user accidentally created two identical indexes.

Reply:
Then one of them will be filtered out since the collation1 covers collation2 if they are equal.

Reply:
This way we may lose good plans. E.g. we may have `IndexScan[a ASC]` and `IndexLookup[a ASC, filter]`, and the second one will be filtered.

Reply:
Addressed in https://github.com/hazelcast/hazelcast/pull/18013/commits/30dc043f41e9d68727328ffc0302413f07d7bb2e
First, exclude only covered full scans leaving the scans with the "longest" collations. Then filter out full scans covered by the lookups (with the same collation).